### ISO Specific Information

Mostly related to how version is represented and rolecode details. 

Are the rolecode URLs still available? How many are from an unexpected source? 

Do any responses reference code lists from different sources? 

How many responses (or host systems) contain a reference to a codelist that's no longer available? 

And what informational content will be affected, ie are the spatial reference blocks more likely to reference a dead code list link?


How are ISO versions represented? Is it clear enough to understand how it might be structured?

(So, no. What matters more for parsing and reading the XML is more keyed to the schema definition. SMAP is interesting in that there's a mimetype (unnecessary) but I think not a custom schema - it's a standardized structure for SMAP in ISO but is just about how to use these elements in this way to reduce the variability. It should just be ISO and validated that way; knowing it's SMAP doesn't affect anything *if* a client is built to the spec, not the server (haha. hahaha.).) But no, the version string gives you a bit of a clue to where some of those other, similar uses of ISO. This is different than INSPIRE, right, that one is an additional namespace so it is validated for ISO and for INSPIRE structure.


In [2]:
%matplotlib inline
import pandas as pd
import json as js  # name conflict with sqla
import sqlalchemy as sqla
from sqlalchemy.orm import sessionmaker
from IPython.display import display

In [3]:
# grab the clean text from the rds
with open('../local/big_rds.conf', 'r') as f:
    conf = js.loads(f.read())

# our connection
engine = sqla.create_engine(conf.get('connection'))

In [ ]:
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_colwidth', 1000):
    display(df[:10])

### Versions

Where to find ISO version information:



| Standard Name                                                       |  Version                                                                | 
|---------------------------------------------------------------------|-------------------------------------------------------------------------| 
| //*/metadataStandardName/CharacterString                            |  //*/metadataStandardVersion/CharacterString                            | 
| //*/seriesMetadata/MD_Metadata/metadataStandardName/CharacterString |  //*/seriesMetadata/MD_Metadata/metadataStandardVersion/CharacterString | 
| //*/seriesMetadata/MI_Metadata/metadataStandardName/CharacterString |  //*/seriesMetadata/MI_Metadata/metadataStandardVersion/CharacterString | 


This is just the version info for single metadata records (not the DS proper, but the MI/MD contained by a DS). [Speculation] This bit of information is not often included in the CSW or OAI-PMH getrecords outputs (those never track as valid ISO and certainly not at all complete ISO compared to what can be accessed from other parts of those larger portal systems). That last is important? or not, depends on whether we think CSW or OAI-PMH is functional (slow, so painfully slow in current implementations). They do not track as first-class dev efforts - must have it but not key to the system so not well-built and not well-maintained. But, back one of the points, did not validate any of those internal ISO records (if we have any and we do) and, as noted previously, the specs for XML validation allow for nested vaildation but that is not implemented. PParse, part of the Apache Xerces package, is (been a while since I had to go through that eval) one of the view available free options and it does not support that kind of validation.

In [7]:
# the set of unique version strings

sql = """
select distinct standard_name, standard_version
from iso_versions
order by standard_name, standard_version;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 75):
    display(df[:61])

standard_name  \
0   ANZLIC Metadata Profile: An Australian/New Zea...   
1                           DM - Regole tecniche RNDT   
2                                              Gemini   
3                                              GEMINI   
4             INSPIRE Implementing Rules for Metadata   
5                                            ISO19115   
6                                            ISO19115   
7                                           ISO 19115   
8                                           ISO 19115   
9                                           ISO 19115   
10                                    ISO 19115/19139   
11                                        ISO 19115-2   
12                               ISO 19115:2003/19139   
13                          ISO 19115/2003/Cor.1:2006   
14                            ISO 19115:2003 (GDI-BW)   
15  ISO 19115:2003 - Geographic information - Meta...   
16  ISO 19115-2 Geographic information - Metadata ...   
17  ISO 19115-2 Geographic information -- Metadata...   
18  ISO 19115-2 Geographic information Metadata Pa...   
19  ISO 19115-2 Geographic information - Metadata ...   
20  ISO 19115-2 Geographic information - Metadata ...   
21  ISO 19115-2 Geographic Information - Metadata ...   
22  ISO 19115-2 Geographic Information - Metadata ...   
23  ISO 19115-2 Geographic Information - Metadata ...   
24  ISO 19115-2 Geographic Information - Metadata ...   
25  ISO 19115-2 Geographic Information - Metadata ...   
26  ISO 19115-2 Geographic information - Metadata ...   
27  ISO 19115-2 Geographic Information - Metadata ...   
28  ISO 19115-2 Geographic Information - North Ame...   
29        ISO 19115 Geographic information - Metadata   
30        ISO 19115 Geographic Information - Metadata   
31        ISO 19115 Geographic Information - Metadata   
32  ISO 19115 Geographic information - Metadata - ...   
33  ISO 19115 Geographic information Metadata; WIS...   
34                       ISO 19115/SeaDataNet profile   
35                              ISO 19115 (UK GEMINI)   
36                                           ISO19119   
37                                          ISO 19119   
38              ISO 19139/19115 Metadata for Datasets   
39          ISO 19139/19119 Metadata for Web Services   
40  ISO 19139 Geographic Information - Metadata - ...   
41                                      ISO-NAP-USGIN   
42                                      ISO-NAP-USGIN   
43                                          ISO-USGIN   
44                  MEDIN Discovery metadata standard   
45                  MEDIN Discovery metadata standard   
46                  MEDIN Discovery metadata standard   
47                  MEDIN Discovery Metadata Standard   
48                  MEDIN Discovery Metadata Standard   
49                  MEDIN Discovery Metadata Standard   
50                  MEDIN Discovery Metadata Standard   
51                                     NAP - Metadata   
52                     NERC profile of ISO:19115:2003   
53                      NERC profile of ISO19115:2003   
54                      NERC profile of ISO19115:2003   
55  North American Profile of ISO 19115:2003 - Geo...   
56  North American Profile of ISO19115:2003 - Geog...   
57  North American Profile of ISO 19115 Geographic...   
58                                          UK GEMINI   
59                                          UK GEMINI   
60                          WMO Core Metadata Profile   

                                     standard_version  
0                                                 1.1  
1                                    10 novembre 2011  
2                                                 2.1  
3                                                 2.1  
4                                                 1.2  
5                                                 1.0  
6                                     2003/Cor.1:2006  
7                                                2003  
8

In [11]:
# let's go with pct of all iso per version string, out of 19,689
sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select count(*) as total
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
	)

	select r.id, a.total
	from responses r join i on i.response_id = r.id
		natural inner join a 
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, count(v.response_id) as num,
	round(count(v.response_id) / max(j.total)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version
order by pct_of_iso DESC;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 55, 'display.max_colwidth', 500):
    display(df[:54])
    
# the sum is 200 shy 
# TODO: verify that count (not from versions unless there're duplicate entries?)

the_version  \
0                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
1                                                                                                       ISO 19115 Geographic Information - Metadata; 2009-02-15   
2                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
3                                                                                                                                     ISO19115; 2003/Cor.1:2006   
4                                                                                                                                                ISO-USGIN; 1.2   
5                                                                                                                                               ISO 19115; 2003   
6                                                                                                                                          ISO-NAP-USGIN; 1.1.4   
7                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
8                                                                                                              MEDIN Discovery Metadata Standard; Version 2.3.5   
9                                                                                                               ISO 19139/19119 Metadata for Web Services; 2005   
10                                                                                                       ISO 19115 Geographic Information - Metadata; ISO 19115   
11                                                                                                                             ISO 19115 (UK GEMINI); 1.0 (2.2)   
12                                                                                               ISO 19115 Geographic information - Metadata; ISO 19115:2003(E)   
13                                                                                                                           NERC profile of ISO19115:2003; 1.0   
14                                         ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
15                                  ISO 19115-2 Geographic Information - North American Profile Metadata - Data with Biological Extensions; ISO 19115-2:2009(E)   
16                                                                                                                                    ISO 19115:2003/19139; 1.0   
17                                                  ISO 19115 Geographic information - Metadata - Converted from Data.gov legacy DMS format.; ISO 19115:2003(E)   
18                                                                                                                     MEDIN Discovery metadata standard; 2.3.4   
19                                                                                                                                 ISO 19115:2003 (GDI-BW); 1.2   
20                                                                             ISO 19139 Geographic Information - Metadata - Implementation Specification; 2007   
21                                                                                                                 INSPIRE Implementing Rules for Metadata; 1.2   
22                                            ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009   
23                                                                                                                                                  GEMINI; 2.1   
24                                                                       

In [ ]:
# pct of the iso responses, ordered by number of responses for the version string,
# per host

sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select r.host, count(r.id) as num_per_host
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
		group by r.host
	)

	select r.id, r.host, a.num_per_host
	from responses r join i on i.response_id = r.id
		join a on a.host = r.host
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, j.host, --count(j.id) as num_iso_responses,
	round(count(j.id) / max(j.num_per_host)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version, j.host
--order by j.host, pct_of_iso DESC;
order by count(v.response_id) DESC;
"""
    
df = pd.read_sql(sql, engine)

In [8]:
with pd.option_context('display.max_rows', 150):
    display(df[:145])

the_version  \
0    ISO 19115-2 Geographic Information - Metadata ...   
1    ISO 19115 Geographic Information - Metadata; 2...   
2                                       ISO-USGIN; 1.2   
3                                      ISO 19115; 2003   
4                                 ISO-NAP-USGIN; 1.1.4   
5    ISO 19115-2 Geographic Information - Metadata ...   
6    ISO 19115-2 Geographic Information - Metadata ...   
7    ISO 19115 Geographic Information - Metadata; 2...   
8    ISO 19115-2 Geographic Information - Metadata ...   
9                            ISO19115; 2003/Cor.1:2006   
10                           ISO19115; 2003/Cor.1:2006   
11     ISO 19139/19119 Metadata for Web Services; 2005   
12    MEDIN Discovery Metadata Standard; Version 2.3.5   
13   ISO 19115-2 Geographic Information - Metadata ...   
14                           ISO19115; 2003/Cor.1:2006   
15   ISO 19115 Geographic Information - Metadata; I...   
16   ISO 19115-2 Geographic Information - Metadata ...   
17   ISO 19115-2 Geographic Information - Metadata ...   
18   ISO 19115-2 Geographic Information - Metadata ...   
19   ISO 19115-2 Geographic Information - Metadata ...   
20                    ISO 19115 (UK GEMINI); 1.0 (2.2)   
21   ISO 19115-2 Geographic Information - Metadata ...   
22   ISO 19115-2 Geographic Information - Metadata ...   
23                  NERC profile of ISO19115:2003; 1.0   
24   ISO 19115-2 Geographic Information - Metadata ...   
25   ISO 19115-2 Geographic Information - Metadata ...   
26                    ISO 19115 (UK GEMINI); 1.0 (2.2)   
27   ISO 19115-2 Geographic Information - Metadata ...   
28   ISO 19115-2 Geographic Information - Metadata ...   
29   ISO 19115-2 Geographic Information - Metadata ...   
30   ISO 19115-2 Geographic Information - Metadata ...   
31   ISO 19115-2 Geographic Information - Metadata ...   
32   ISO 19115 Geographic information - Metadata; I...   
33                           ISO 19115:2003/19139; 1.0   
34   ISO 19115-2 Geographic Information - North Ame...   
35   ISO 19115-2 Geographic Information - Metadata ...   
36   ISO 19115-2 Geographic information - Metadata ...   
37   ISO 19115 Geographic information - Metadata; I...   
38   ISO 19115 Geographic Information - Metadata; I...   
39   ISO 19115 Geographic information - Metadata - ...   
40   ISO 19115-2 Geographic information - Metadata ...   
41                        ISO 19115:2003 (GDI-BW); 1.2   
42                                      ISO-USGIN; 1.2   
43            MEDIN Discovery metadata standard; 2.3.4   
44   ISO 19115-2 Geographic Information - Metadata ...   
45   ISO 19115-2 Geographic Information - Metadata ...   
46                           ISO19115; 2003/Cor.1:2006   
47        INSPIRE Implementing Rules for Metadata; 1.2   
48   ISO 19139 Geographic Information - Metadata - ...   
49    MEDIN Discovery Metadata Standard; Version 2.3.5   
50   ISO 19115 Geographic information - Metadata; I...   
51            MEDIN Discovery metadata standard; 2.3.4   
52                                         GEMINI; 2.1   
53            MEDIN Discovery metadata standard; 2.3.8   
54                                     ISO 19115; 2003   
55   ISO 19115-2 Geographic information - Metadata ...   
56   ISO 19115-2 Geographic Information - Metadata ...   
57   ANZLIC Metadata Profile: An Australian/New Zea...   
58   ISO 19115-2 Geographic Information - Metadata ...   
59   ISO 19115 Geographic information - Metadata - ...   
60              WMO Core Metadata Profile; 1.0 (draft)   
61   ISO 19115-2 Geographic Information - North Ame...   
62                           ISO19115; 2003/Cor.1:2006   
63                     ISO 19115/19139; 2003/cor1:2006   
64   ISO 19115-2 Geographic Information - Metadata ...   
65   ISO 19115-2 Geographic Information - Metadata ...   
66                                      UK GEMINI; 2.2   
67   ISO 19115-2 Geographic Information - Metadata ...   
68                  NERC profile of ISO19

In [12]:
sql = """
with j as
(
	with i as
	(
		select d.response_id, jsonb_array_elements(d.identity::jsonb) ident
		from identities d
		where d.identity is not null
	), a as (
		select r.host, count(r.id) as num_per_host
		from responses r join i on i.response_id = r.id
		where i.ident->>'protocol' = 'ISO'
		group by r.host
	)

	select r.id, r.host, a.num_per_host
	from responses r join i on i.response_id = r.id
		join a on a.host = r.host
	where i.ident->>'protocol' = 'ISO'
), v as 
(
	select response_id,
		standard_name || '; ' || standard_version as the_version
	from iso_versions
)

select v.the_version, j.host, --count(j.id) as num_iso_responses,
	round(count(j.id) / max(j.num_per_host)::numeric * 100.0, 2) as pct_of_iso
from responses r join v on v.response_id = r.id
	join j on j.id = r.id
group by v.the_version, j.host
--order by j.host, pct_of_iso DESC;
order by j.host, pct_of_iso DESC;
"""
    
df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 150, 'display.max_colwidth', 400):
    display(df[:145])

the_version  \
0                                                                                                                               ISO 19115-2; ISO 19115-2:2009(E)   
1                                              ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
2                                           ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
3                                                                                                        ISO 19115 Geographic Information - Metadata; 2009-02-15   
4                                                                                                                                      ISO 19115:2003/19139; 1.0   
5                                                                                                 ISO 19115 Geographic information - Metadata; ISO 19115:2003(E)   
6                                                                                                         ISO 19115 Geographic Information - Metadata; ISO 19115   
7                                                                                                                                                 ISO-USGIN; 1.2   
8                                              ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
9                                                    ISO 19115 Geographic information - Metadata - Converted from Data.gov legacy DMS format.; ISO 19115:2003(E)   
10                                                                                            ISO 19115:2003 - Geographic information - Metadata; ISO 19115:2003   
11                                          ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
12                                                                                                                   ISO 19139/19115 Metadata for Datasets; 2003   
13                                                                                                                                                 ISO19115; 1.0   
14                                                                                                                                            ISO-NAP-USGIN; 1.1   
15                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2013(E)   
16                                             ISO 19115-2 Geographic information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009   
17                                                                              ISO 19139 Geographic Information - Metadata - Implementation Specification; 2007   
18                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
19                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for imagery and gridded data; ISO 19115-2:2009(E)   
20                                                                                                                                     ISO19115; 2003/Cor.1:2006   
21                                          ISO 19115-2 Geographic Information - Metadata - Part 2: Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
22                                             ISO 19115-2 Geographic Information - Metadata Part 2 Extensions for Imagery and Gridded Data; ISO 19115-2:2009(E)   
23                                                                                                                                     ISO 19115:2003/19139; 1.0   
24                                               

### Codelists


In [13]:
# how many unique codelist references (before the hash only)

sql = """
with a as
(
	select replace(left(codelist, strpos(codelist, '#')), '#', '') as codelist
	from codelists
)
select distinct codelist
from a
where codelist != ''
order by codelist;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 60, 'display.max_colwidth', 400):
    display(df[:52])

codelist
0                                               file://Agi-s000001/afdelingen/GSMS/Extern/Applicaties_en_Tools/Databeheer/Metadata/GeoMDE_22/resources/customCodelists_new.xml
1                                                                      file://appconf.intranet.rws.nl/Appconf/metadatatools/Metadatamaker/config_bestanden/customCodelists.xml
2                                                                                      file://Shr-ipvw-gpr001/metadatatools/Metadatamaker/config_bestanden/customCodelists.xml
3                                                                                                          http://adl.brs.gov.au/anrdl/resources/codeList/codeList20120313.xml
4                                                                                                                   http://asdd.ga.gov.au/asdd/profileinfo/GAScopeCodeList.xml
5                                                                                                                      http://asdd.ga.gov.au/asdd/profileinfo/gmxCodelists.xml
6                                                                                     http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/Codelist/gmxCodelists.xml
7                                                                                  http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/Codelist/ML_gmxCodelists.xml
8                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/CARIS_20120814/resources/Codelist/gmxCodelists.xml
9                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/CARIS_20120814/resources/Codelist/srvCodelists.xml
10                                                                                                       http://data.daff.gov.au/anrdl/resources/codeList/codeList20120313.xml
11                                                                                                             http://geobrain.laits.gmu.edu/CatalogXML/resources/codeList.xml
12                                                                                                                      http://mdtranslator.adiwg.org/api/codelists?format=xml
13                                                                                                        http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml
14                                                                                           http://schemas.opengis.net/iso/19139/20070417/resources/Codelist/gmxCodelists.xml
15                                                                                                             http://seadata.bsh.de/isoCodelists/sdnCodelists/csrCodeList.xml
16                                                                                             http://seadatanet.maris2.nl/isocodelists/sdncodelists/edmo-edmerp-Codelists.xml
17                                                                                                      http://standards.iso.org/iso/19139/resources/codelist/gmxCodelists.xml
18                                                              http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/gmxCodelists.xml
19                                                              http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/Codelist/gmxCodelists.xml
20                                                           http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/ML_gmxCodelists.xml
21                                                          http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/Codelist/ML_gmx Codelists.xml
22                                                           http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resou

In [18]:
# how many of those are http or https links?
# AND some are casing diffs only, so lowercase all
sql = """
with a as
(
	select lower(replace(left(codelist, strpos(codelist, '#')), '#', '')) as codelist
	from codelists
)
select distinct codelist
from a
where codelist ilike 'http%%'
order by codelist;
"""

df = pd.read_sql(sql, engine)
with pd.option_context('display.max_rows', 60, 'display.max_colwidth', 400):
    display(df[:46])

codelist
0                                                                                                          http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml
1                                                                                                                   http://asdd.ga.gov.au/asdd/profileinfo/gascopecodelist.xml
2                                                                                                                      http://asdd.ga.gov.au/asdd/profileinfo/gmxcodelists.xml
3                                                                                     http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/gmxcodelists.xml
4                                                                                  http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/ml_gmxcodelists.xml
5                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/gmxcodelists.xml
6                                                                               http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/srvcodelists.xml
7                                                                                                        http://data.daff.gov.au/anrdl/resources/codelist/codelist20120313.xml
8                                                                                                              http://geobrain.laits.gmu.edu/catalogxml/resources/codelist.xml
9                                                                                                                       http://mdtranslator.adiwg.org/api/codelists?format=xml
10                                                                                                        http://nap.geogratis.gc.ca/metadata/register/napmetadataregister.xml
11                                                                                           http://schemas.opengis.net/iso/19139/20070417/resources/codelist/gmxcodelists.xml
12                                                                                                             http://seadata.bsh.de/isocodelists/sdncodelists/csrcodelist.xml
13                                                                                             http://seadatanet.maris2.nl/isocodelists/sdncodelists/edmo-edmerp-codelists.xml
14                                                                                                      http://standards.iso.org/iso/19139/resources/codelist/gmxcodelists.xml
15                                                              http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/gmxcodelists.xml
16                                                          http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmx codelists.xml
17                                                           http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmxcodelists.xml
18                                                           http://standards.iso.org/ittf/publ.*lyavailablestandards/iso_19139_schemas/resources/codelist/ml_gmxcodelists.xml
19                                                                                                        http://vocab.nerc.ac.uk/isocodelists/sdncodelists/cdicsrcodelist.xml
20                                                                                                         http://vocab.nerc.ac.uk/isocodelists/sdncodelists/edmedcodelist.xml
21                                                                                                          http://vocab.nerc.ac.uk/isocodelists/sdncodelists/gmxcodelists.xml
22     http://webarchive.nationalarchives.gov.uk/20130402145952/http://standards.iso.org/ittf/publiclyavailablestandards/iso_19139_schemas/re

In [20]:
# and a super quick linkrot check
import requests

codes = []
for d in df.itertuples():
    url = d[1]
    try:
        rsp = requests.get(url, timeout=30)
        status_code = rsp.status_code
    except Exception as ex:
        print url, ex
        status_code = 900
    
    codes.append((url, status_code))
    
    
    

http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml ('Connection aborted.', gaierror(8, 'nodename nor servname provided, or not known'))
http://www.ec.gc.ca/data_donnees/standards/schemas/napec ('Connection aborted.', error(54, 'Connection reset by peer'))
http://www.tc211.org/iso19139/resources/codelist.xml ('Connection aborted.', gaierror(8, 'nodename nor servname provided, or not known'))


In [22]:
from collections import defaultdict
statuses = defaultdict(int)
for url, code in codes:
    print code, url
    statuses[code] += 1

900 http://adl.brs.gov.au/anrdl/resources/codelist/codelist20120313.xml
404 http://asdd.ga.gov.au/asdd/profileinfo/gascopecodelist.xml
404 http://asdd.ga.gov.au/asdd/profileinfo/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/20070417/resources/codelist/ml_gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/gmxcodelists.xml
200 http://aws2.caris.com/sfs/schemas/iso/19139/caris_20120814/resources/codelist/srvcodelists.xml
404 http://data.daff.gov.au/anrdl/resources/codelist/codelist20120313.xml
404 http://geobrain.laits.gmu.edu/catalogxml/resources/codelist.xml
200 http://mdtranslator.adiwg.org/api/codelists?format=xml
404 http://nap.geogratis.gc.ca/metadata/register/napmetadataregister.xml
404 http://schemas.opengis.net/iso/19139/20070417/resources/codelist/gmxcodelists.xml
404 http://seadata.bsh.de/isocodelists/sdncodelists/csrcod

In [23]:
statuses

defaultdict(<type 'int'>, {200: 14, 404: 25, 900: 3})

In [24]:
len(codes)

42

In [28]:
for k, v in statuses.iteritems():
    print k, '{:.2f}%'.format(v / float(len(codes)) * 100.)

200 33.33%
404 59.52%
900 7.14%


####So how many responses contain an inaccessible codelist?

In [31]:
sql = """
select count(distinct r.response_id)
from response_codelists r join codelist_statuses c on c.codelist = r.codelist
where c.status != 200;
"""
df = pd.read_sql(sql, engine)
df

count
0  16129

So 16,129 out of 19,689 ISO records have at least one reference to a downed codelist. 82%.

Question: Do you want to know what the downed codelists affect (role codes or progress codes or ??)? 